File: cpp/0378.0_Kth_Smallest_Element_in_a_Sorted_Matrix.cpp
/*
执行用时：20 ms, 在所有 C++ 提交中击败了92.92% 的用户
内存消耗：14.4 MB, 在所有 C++ 提交中击败了8.26% 的用户
通过测试用例：86 / 86
*/

class Solution {
public:
    int kthSmallest(vector<vector<int>>& matrix, int k) {
        priority_queue<int> q;
        for (auto row: matrix) {
            for (auto num: row) {
                q.push(num);
                if (q.size() > k) {
                    q.pop();
                }
            }
        }
        return q.top();
    }
};

File: problems/0378.0_Kth_Smallest_Element_in_a_Sorted_Matrix.py
'''
binary search
T: O(2 * N * log(max(*matrix) - min(*matrix)))
S: O(1)

Runtime: 272 ms, faster than 65.16% of Python3 online submissions for Kth Smallest Element in a Sorted Matrix.
Memory Usage: 18.8 MB, less than 38.85% of Python3 online submissions for Kth Smallest Element in a Sorted Matrix.
'''
class Solution:
    def kthSmallest(self, matrix: List[List[int]], k: int) -> int:
        n = len(matrix)
        lower, upper = matrix[0][0], matrix[-1][-1]
        ans = lower 
        
        def count_le(x):
            '''count less than or equal to x, sum of every row count'''
            cnt = 0
            i, j = 0, n - 1
            while i < n and j >= 0:
                if matrix[i][j] > x:
                    j -= 1
                else:
                    cnt += j + 1
                    i += 1
            return cnt 
        
        while lower <= upper:
            mid = (lower + upper) // 2 
            if count_le(mid) >= k:
                ans = mid 
                upper = mid - 1
            else:
                lower = mid + 1
        return ans 
